core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java

```diff
@@ -251,10 +251,10 @@ public void start(WorkerNetAddress address) throws IOException {
    */
   @Override
   public void stop() throws IOException {
-    // Stop heart-beat executors and clients.
-    mResourceCloser.close();
     // Stop the base. (closes executors.)
     super.stop();
+    // Stop heart-beat executors and clients.
```

Comment:
It seems like this order is important. Do we know why this has to be the order, and can we add that into the comments here? It is not obvious with the code.

Reply:
I'm still trying to wrap my head around _why_ this even works. Based on the code path the threads don't seem to be daemon threads, and the code path of the heartbeat threads don't indicate anywhere that they respond to thread interrupts.

I will add a comment once I understand why

Reply:
Do you have a consistent repro for this? 
I'd guess we'd still hit this issue without `AbstractClient` bailing out on interrupt.

Reply:
Looks like `BlockMasterSync`'s heartbeat could be stuck if close here is not reversed?

Reply:
I have not been able to repro it after only making this change.

Without this PR, the consistent repro is

```console
$ ./bin/alluxio-start.sh master & ./bin/alluxio-start.sh worker & wait
$ ./bin/alluxio-stop.sh master
$ ./bin/alluxio-stop.sh worker
```

If you want to grab a jstack during that time you can run:

```console
$ ./bin/alluxio-start.sh master & ./bin/alluxio-start.sh worker & wait
$ ./bin/alluxio-stop.sh master && sleep 5 && (./bin/alluxio-stop.sh worker & sleep 1 && jstack -l `jps | grep AlluxioWorker | awk '{print $1}'`)
```

Reply:
This looks very aggressive, in restarting the master and that strengthens my hypothesis of `BlockMasterSync` being the culprit. After all, it's the only one that actually closes something in its `#close()`

Reply:
Looks like this is caused by sharing `DefaultBlockWorkerClient::mFileSystemMasterClient` with `PinListSync`.

`AbstractClient`'s `close` and `connect` methods are synchronized. While `PinListSync` is trying to connect (to a master that is down), `DefaultBlockWorker` is trying to close the very same client as part of `mResourceCloser.close()` call. So shutdown will be blocked until the `PinListSync` times out.

It seems, shutting down the base-executor first bails `PinListSync` out and later call to `mResourceCloser.close()` doesn't get stuck. This means we don't need that bail-out condition on `AbstractClient` as it seems to handle interruptions somewhere else.

That being said, I believe the proper fix is to host `FileSystemClientPool` in `DefaultBlockWorker` instead of a single `FileSystemMasterClient`. So we can isolate `FileSystemMasterClient` resources between the `DefaultBlockWorker` and the `PinListSync`.

Reply:
@ggezer I implemented your suggested changes in the previous comment by creating an FS client pool instead. While at this point clients weren't locked on each other. It seemed that they now _all_ hang trying to connect to the master.

https://pastebin.com/Nc6yVr2q


The worker block sync still blocked shutdown:

Reply:
I see. Closing the pools is blocked.

> "alluxio-process-shutdown-hook" #292 prio=5 os_prio=0 tid=0x00007f6e2c002800 nid=0xdd9f waiting for monitor entry [0x00007f724f8f8000]
>    java.lang.Thread.State: BLOCKED (on object monitor)
>     at alluxio.AbstractClient.close(AbstractClient.java:310)
>     - waiting to lock <0x00000006c04cf6a8> (a alluxio.worker.block.BlockMasterClient)
>     at com.google.common.io.Closer.close(Closer.java:214)
>     at alluxio.worker.block.BlockMasterClientPool.close(BlockMasterClientPool.java:55)
>     at com.google.common.io.Closer.close(Closer.java:214)
>     at alluxio.worker.block.DefaultBlockWorker.stop(DefaultBlockWorker.java:261)
>     at alluxio.Registry.stop(Registry.java:148)

Reply:
Even though it's quite disruptive I'm OK to swap the lines at this point as I can't see a way around `HeartbeatExecutor`'s limitations.

Reply:
Ideally, we'd like to implement your initial commit as an abstract utility to all heartbeat executors at some point.

Reply:
@ZacBlanco Do we know why this order works, and the other order didn't?

Reply:
I believe this ordering works because `super.stop` will issue a `shutdownNow` on the executor which automatically sends an interrupt to all executor threads.

With this ordering, no interrupts are sent and we try to manually stop/close the heartbeat threads. Since no interrupt was performed, the heartbeat thread continues trying to connect until timing out.

Reply:
@gpang without first interrupting the threads, `mResourceCloser.close()` gets blocked because `AbstractClient.close()` is synchronized with `AbstractClient.connect()` and `connect` was trying to connect until it times out.

Reply:
Can we add this info as a comment, since this ordering is intentional and critical.

Reply:
updated